homework-5/homework-5.py

```diff
@@ -0,0 +1,86 @@
+# Backend Programming with Python
+# Homework 5 - Battleships
+
+import random, math
+
+
+class SpaceShip:
+    def __init__(self, name, hp, armour, damage):
+        self.name = name
+        self.hp = hp
+        self.armour = armour
+        self.damage = damage
+
+    def attack(self, other):
+        if isinstance(other, SpaceShip):
+            damage_dealt = self.damage - other.armour
+            if damage_dealt < 0 < self.damage:
+                damage_dealt = 1
+            other.hp = other.hp - damage_dealt
+        print(
+            f'Damage dealt to {other.name} = {self.damage} ({self.name} Damage) - {other.armour} ({other.name} Armor) = {damage_dealt}')
+        print(f'HP left on {other.name} = {other.hp}')
+
+    def upgrade(self, upgrade_name):
```

Comment:
It's a nice try at the extra and I heartily appreciate the effort, but unfortunately it won't work, sorry! The problems appear when you start combining upgrades:
1. You can't check if an upgrade has already been added  

2. The calculations are incorrect when you look at the requirements.  
3. Let's assume the following upgrades on a 2 Armor ship (added in this order):
** Upgrade 1: +3 Armor
** Upgrade 2: +50% Armor
** Upgrade 3: +5 Armor
** Upgrade 4: +50% Armor

How it should work: 2 + (2 x (50% + 50%)) + (3 + 5) = 2 + 2 + 8 = 12
How it works for you: (((2 + 3) x 150%) + 5) x 150% = (7 + 5) x 150% = 12 x 150% = 18